cylc/flow/cfgspec/suite.py

```diff
@@ -335,9 +344,135 @@ def upg(cfg, descr):
             )
+
     except KeyError:
         pass
 
+    # TODO - uncomment this fn so that we actually use the host to platform
+    # upgrader
+    # cfg = host_to_platform(cfg)
+
+
+def host_to_platform_upgrader(cfg):
+    """Upgrade a config with host settings to a config with platform settings
+    if it is appropriate to do so.
+
+                       +-------------------------------+
+                       | Is platforms set in this      |
+                       | [runtime][TASK]?              |
+                       +-------------------------------+
+                          |YES                      |NO
+                          |                         |
+    +---------------------v---------+      +--------+--------------+
+    | Are any forbidden items set   |      | host == $(function)?  |
+    | in any [runtime][TASK]        |      +-+---------------------+
+    | [job] or [remote] section     |     NO |          |YES
+    |                               |        |  +-------v------------------+
+    +-------------------------------+        |  | Log - evaluate at task   |
+              |YES            |NO            |  | submit                   |
+              |               +-------+      |  |                          |
+              |                       |      |  +--------------------------+
+    +---------v---------------------+ |      |
+    | Fail Loudly                   | |    +-v-----------------------------+
+    +-------------------------------+ |    | * Run reverse_lookup()        |
+                                      |    | * handle reverse lookup fail  |
+                                      |    | * add platform                |
+                                      |    | * delete forbidden settings   |
+                                      |    +-------------------------------+
+                                      |
+                                      |    +-------------------------------+
+                                      +----> Return without changes        |
+                                           +-------------------------------+
+
+    Args (cfg):
+        config object to be upgraded
+
+    Returns (cfg):
+        upgraded config object
+    """
+    # If platform and old settings are set fail
+    # and remote should be added to this forbidden list
+    forbidden_with_platform = [
+        'host', 'batch system', 'batch submit command template'
+    ]
+
+    for task_name, task_spec in cfg['runtime'].items():
+        # if task_name == 'delta':
+        #     breakpoint(header=f"task_name = {task_name}")
+
+        if (
+            'platform' in task_spec and 'job' in task_spec or
+            'platform' in task_spec and 'remote' in task_spec
+        ):
+            if (
+                'platform' in task_spec and
+                any(forbidden_with_platform == task_spec['job']) or
+                any(forbidden_with_platform == task_spec['remote'])
```

Comment:
These `any` comparisons don't look right, should they be set comparisons:

```
forbidden_with_platform = { 'host', ... }
# ...
forbidden_with_platform & {*task_spec['job'], *task_spec['remote']}
```

Reply:
Well for one thing this comment made me realise that I wasn't testing this logic!